src/frontend/src/flows/addDevice/welcomeView/showVerificationCode.ts

```diff
@@ -0,0 +1,131 @@
+    <div id="verificationCode" class="highlightBox">
+      ${tentativeRegistrationInfo.verification_code}
+    </div>
+    <p>Time remaining: <span id="timer"></span></p>
+    <button id="showCodeCancel">Cancel</button>
+  </div>
+`;
+
+/**
+ * Page to display the verification code which is received after successfully registering a tentative device.
+ * @param userNumber Anchor the device to be verified belongs to
+ * @param alias Alias of the tentative device
+ * @param tentativeRegistrationInfo Verification code and timeout received when registering the tentative device
+ * @param credentialToBeVerified Credential id of the device to be verified. When this id appears in the list of authenticators, verification was successful.
+ */
+export const showVerificationCode = async (
+  userNumber: bigint,
+  alias: string,
+  tentativeRegistrationInfo: TentativeRegistrationInfo,
+  credentialToBeVerified: CredentialId
+): Promise<void> => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(userNumber, alias, tentativeRegistrationInfo), container);
+  return init(
+    userNumber,
+    tentativeRegistrationInfo.device_registration_timeout,
+    credentialToBeVerified
+  );
+};
+
+function startPolling(
+  userNumber: bigint,
+  credentialToBeVerified: Array<number>,
+  shouldStop: () => boolean
+): Promise<boolean> {
+  return IIConnection.lookupAuthenticators(userNumber).then((deviceData) => {
+    if (shouldStop()) {
+      return false;
+    }
+    for (const device of deviceData) {
+      if (device.credential_id.length === 1) {
+        const credentialId = device.credential_id[0];
+
+        if (credentialIdEqual(credentialId, credentialToBeVerified)) {
+          return true;
+        }
+      }
+    }
+    return startPolling(userNumber, credentialToBeVerified, shouldStop);
+  });
+}
+
+const init = async (
+  userNumber: bigint,
+  endTimestamp: bigint,
+  credentialToBeVerified: CredentialId
+): Promise<void> => {
+  const countdown = setupCountdown(
+    endTimestamp,
+    document.getElementById("timer") as HTMLElement,
+    async () => {
+      await displayError({
+        title: "Timeout Reached",
+        message:
+          'The timeout has been reached. For security reasons the "add device" process has been aborted.',
+        primaryButton: "Ok",
+      });
+      // TODO L2-309: do this without reload
+      window.location.reload();
+    }
+  );
+  startPolling(userNumber, credentialToBeVerified, countdown.isStopped).then(
+    (verified) => {
+      if (verified) {
+        setUserNumber(userNumber);
+        // TODO L2-309: do this without reload
+        window.location.reload();
+      }
+    }
+  );
+
+  const cancelButton = document.getElementById(
+    "showCodeCancel"
+  ) as HTMLButtonElement;
+
+  cancelButton.onclick = () => {
+    countdown.stop();
+    // TODO L2-309: do this without reload
+    window.location.reload();
+  };
+};
+
+function credentialIdEqual(
```

Comment:
note: the [`sign()` function in `multiWebAuthnIdentity.ts`](https://github.com/dfinity/internet-identity/blob/2ae97d1a79b991f7b47aca4b24c8dea6149bb235/src/frontend/src/utils/multiWebAuthnIdentity.ts#L88) does something similar, albeit with a different implementation. Not saying they _should_ be the same, just thought you might want to have a look.

Reply:
Hm, I think this implementation is more straight forward than converting to buffer first. I'll keep it as is for now.